PyScript Core Documentation (#1560)
pyscript/pyscript: pyscript.core/docs/README.md

 * [micropython](https://micropython.org/) is the name of the interpreter that runs a small subset of the *Python* standard library and is optimized to run in constrained environments such as *Mobile* phones, or even *Desktop*, thanks to its tiny size and an extremely fast bootstrap
 * [wasmoon](https://github.com/ceifa/wasmoon) is the name of the interpreter that runs *Lua* on the browser and that, among the previous two interpreters, is fully compatible with all core features
 * [ruby-wasm-wasi](https://github.com/ruby/ruby.wasm) is the name of the (currently *experimental*) interpreter that adds *Ruby* to the list of programming languages currently supported

While internally the name `interpreter` is used to declare the identity of the reference, this value *must* be specified as `<script type>` attribute:

```html
<script type="pyodide">
    import sys
    print(sys.version)
</script>

<script type="micropython">
    import sys
    print(sys.version)
</script>

<script type="wasmoon">
    print(_VERSION)
</script>

<script type="ruby-wasm-wasi">
    print "ruby #{ RUBY_VERSION }"
</script>
```

> ℹ️ - Please note we decided on purpose to not use the generic programming language name instead of its interpreter project name to avoid being too exclusive for alternative projects that would like to target that very same *PL* (i.e. note *pyodide* & *micropython* not using *python* indeed as interpreter name).

Reviewer: Programming Language most probably

> Last, but not least, it is possible regardless to create any *custom type* using any available interpreter, like we already do with the `<script type="py">` and its `<py-script>` custom element counter-part, both defined through the *@pyscript/core* API. In this case, the `type` won't refer anymore to the *interpreter* but to the *custom type* created in the module's registry.

Author: That's correct because we provide only 4 interpreters that must be pre-bundle and made available through our artifacts. A custom type builds up on top of that interpreter so I think your change makes sense.

  </div>
</details>

<details>

>   <summary><strong>Target</strong></summary>

Author: We do wtih `display` though in PyScript extend ... but that part hasn't been written yet so maybe you are right.

    <script-micropython>Hello</script-micropython>
-->
```

  </div>
</details>

<details>
  <summary><strong>Env</strong></summary>
  <div>

ℹ️ - This is an **advanced feature** that is worth describing but usually it is not needed for most common use cases.

Mostly due its terseness that plays nicely as attribute's suffix, among its commonly understood meaning, we consider an *env* an identifier that guarantee the used *interpreter* would always be the same and no other interpreters, even if they point at very same project, could interfere with globals, behavior, or what's not.

In few words, every single *env* would spawn a new interpreter dedicated to such env, and global variables defined elsewhere will not affect this "_environment_" and vice-versa, an *env* cannot dictate what will happen to other interpreters.

```html
<!-- default env per each interpreter -->
<script type="micropython">
    shared = True
</script>
<script type="micropython">
    # prints True - shared is global
    print(shared)
</script>

<!-- dedicated interpreter -->
<script type="micropython" env="my-project-env">
    # throws an error - shared doesn't exist
    print(shared)

> </script>

Reviewer 2: 👆🏼

</script>
```

Please keep reading this document to understand how to use those utilities or how to have other *Pogramming Languages* enabled in your page via `<script>` elements.

## How Scripts Work

The [&lt;script&gt; element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script) has at least these extremely important peculiarities compared to any other element defined by the [HTML Standard](https://html.spec.whatwg.org/multipage/):

 * its only purpose is to contain *data blocks*, meaning that browsers will never try to parse its content as generic *HTML* (and browsers will completely ignore either its content or its attributes, including the `src`, when its [type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type) is not known)
 * its completely unobtrusive when it comes to both *aria* and *layout*, indeed it's one of the few nodes that can be declared almost anywhere without breaking its parent tree (other notable exception would be a comment node)
 * for our specific use case, it already offers attributes that are historically well understood and known, also simplifying somehow the creation of this document

> The long story short is that any `<script type="non-standard-type">` has zero issues with any browser of choice, but it's true that using some specific *custom type* might lead to future issues in case that `type` could have some special meaning for the future of the Web.
> 
> We encourage everyone to be careful when using this *core* API as we definitively don't want to clash or conflict, by any mean, with what the Web might need or offer in the near to far future, but we're also confident so far our current *types* are more than safe.

Reviewer 3: ```suggestion
While`<script type="non-standard-type">` works with any browser, it's possible that some of the currently non-standard-types become standard in the future. For this reason, we've chosen to use specific names which we feel are safe (i.e. `<script type="pyodide">` instead of `type="python"`) - users are encouraged to use the same caution when specifying their own type names to avoid conflicts with future standards.
```

Reviewer 2: Probably worth mentioning that we "explicitly take that into account when adding default types in core and any plugin adding aliases or support for new `types` should that that into consideration as well", or something like that

| env       | `<script type="pyodide" env="brand">`         | Create, if not known yet, a dedicated *environment* for the specified `type`. Please read the [Terminology](#terminology) **env** dedicated details to know more. |
| src       | `<script type="pyodide" src="./app.py">`      | Fetch code from the specified `src` file, overriding or ignoring the content of the `<script>` itself, if any. |
| target    | `<script type="pyodide" target="outcome">`    | Describe as *id* or *CSS* selector the default *target* to use as `document.currentScript.target` field. Please read the [Terminology](#terminology) **target** dedicated details to know more. |
| type      | `<script type="micropython">`                 | Define the *interpreter* to use with this script. Please read the [Terminology](#terminology) **interpreter** dedicated details to know more. |

> | version   | `<script type="pyodide" version="0.23.2">`    | Allow the usage of a specific version where, if numeric, must be available through the project *CDN* used by *core* but if specified as fully qualified *URL*, allows usage of any interpreter's version: `<script type="pyodide" version="http://localhost:8080/pyodide.local.mjs">` |

Reviewer 3: ```suggestion
| version   | `<script type="pyodide" version="0.23.2">`    |  Load a specific version of the specified interpreter type. If a numeric version is provided (i.e. `0.23.2`) , that version of the interpreter will be loaded from the interpreter's default CDN. If a fully qualified URL is provided, the interpreter will be loaded from that URL.
```

### Script Features

> These are all special, *script* related features, offered by *@pyscript/core* out of the box.

Reviewer 3: ```suggestion
*@pyscript/core* adds the following additional features to all the interpreters it includes:
```

<details>
  <summary><strong>document.currentScript</strong></summary>
  <div>

> No matter the interpreter of choice, if there is any way to reach the `document` from such interpreter, its `currentScript` will point at the exact/very-same script that is currently executing the code, even if its `async` attribute is used, mimicking what the standard [document.currentScript](https://developer.mozilla.org/en-US/docs/Web/API/Document/currentScript) offers already, and in an unobtrusive way for the rest of the page, as this property only exists for *synchronous* and blocking scripts that are running, hence never interfering with this *core* logic or vice-versa.

Reviewer 3: ```suggestion
If the interpreter has a way to access the [document](https://developer.mozilla.org/en-US/docs/Web/API/Document), the document's `currentScript` attribute will contain a reference to the script tag that is currently executing code (even if its `async` attribute is used). This mimics what the standard [document.currentScript](https://developer.mozilla.org/en-US/docs/Web/API/Document/currentScript) offers already. This property only exists for *synchronous* and blocking scripts that are running, and never interferes with *core* logic or vice-versa.
```

Reviewer 2: +1 on the suggestion, especially specifying that it's the script **tag** as it can be easily misinterpreted with the "script code" itself.

```html
<script type="micropython" id="my-target">
    from js import document

    # explicitly grab the current script as target
    my_target = document.getElementById('my-target')

    # verify it is the exact same node with same id
    print(document.currentScript.id == my_target.id)
</script>
```

> 

Reviewer 3: This is where I'd move the description of the `target` property from above - this is really the context where I think it's useful.

Not only this is helpful to crawl the surrounding *DOM* or *HTML*, every script will also have a `target` property that will point either to the element reachable through the `target` attribute, or it lazily creates once a companion element that will be appended right after the currently executing *script*.

> Please read the [Terminology](#terminology) **target** dedicated details to know more.

Reviewer 3: See above on where I think `target` should go.

  </div>
</details>

<details>
  <summary><strong>XWorker</strong></summary>
  <div>

> With or without access to the `document`, every (*non experimental*) interpreter will have defined, at the global level, a reference to the `XWorker` "_class_" (it's just a *function*!), which goal is to enable off-loading heavy operations on a worker, without blocking the main / UI thread (the current page) and allowing such worker to even reach the `document` or anything else available on the very same main / UI thread.

Reviewer 3: I think this is the first reference to experimental vs non-experimental interpreters. Do we intend to release any interpreters that are still experimental? If not, I don't think we need to mention them here.

Reviewer 2: Mmmm... this is the Tech Preview docs. I assume we'll re-work them and change as we work on the next task: "Moving `next` as new core and make the right architecture and high level API changes". The XWorker interface, for instance might change... or we may decide to move the experimental types out of core and make them available as plugins out of core? 

Just want to make sure we are all in agreement that this is not the final/set in stone, architecture/API.

Reviewer 2: I'd also rewrite this in a way to put more emphasis in the `XWorker` API itself rather than the interpreter. Maybe something like:

"
The goal of the`XWorker` API is to enable off-loading heavy operations to a worker, without blocking the main / UI thread (the current page), while also allowing such worker to even reach the `document` or anything else available on the very same main / UI thread.

Note that every (*non experimental*) interpreter will have defined, at the global level, a reference to the `XWorker`, with or without access to the `document`.
"

```html
<script type="micropython">
    # XWorker is globally defined
    print(XWorker != None)
</script>
```

Please read the [XWorker](#xworker) dedicated section to know more.

  </div>
</details>

## How Events Work

Inspired by the current [HTML Standard](https://html.spec.whatwg.org/multipage/webappapis.html#event-handlers):

> the event handler is exposed through a name, which is a string that always starts with "_on_" and is followed by the name of the event for which the handler is intended.

We took a similar approach, replacing that `on` prefix with whatever *interpreter* or *custom type* is available on the page, plus a *dash* `-` to avoid clashing with standards:

```html
<script type="micropython">
    def print_type(event, double):
        # logs "click 4"
        print(f"{event.type} {double(2)}")
</script>
<button micropython-click="print_type(event, lambda x: x * 2)">
    print type
</button>
```

> If this example felt a bit verbose, be ensured custom types would work the same:

Reviewer 3: ```suggestion
The prefix for event handlers can be the full name of an interpreter type, as above, or a prefix associated with a custom type alias/extension:
```

```html
<!-- ℹ️ - requires py-script custom type -->
<button py-click="print(event.type)">
    print type
</button>
```

> What is important to understand about *events* in PyScript is that the text within the attribute is executed just like any other inline or external content is, through the very same *interpreter*, with the notably extra feature that the `event` reference is made temporarily available as *global* by *core*.

Reviewer 3: ```suggestion
The text within the attribute is executed just like any other inline or external content is in a \<script\> tag with the matching type, and through the same *interpreter*. The only difference is that, within code that is triggered by an event, an temporary global [Event object](https://developer.mozilla.org/en-US/docs/Web/API/Event) called `event` is availble.
```

> This really reflects how otherwise native Web inline events handlers work and we think it's a great feature to support ... *but*:

Reviewer 3: ```suggestion
This is the same behavior as native Web inline events handlers.

Some key takeaways of this approach:
```

>  * if your script runs *asynchronously* the `event` might be gone on the main / UI thread and by that time any of its `event.stopPropagation()` or `event.preventDefault()` goodness will be problematic, as too late to be executed

Reviewer 3: *JG Comment: Did we move away from the idea of passing the event object directly, via the object passed by the JS event listener? Would that solve this issue with asynchronous handlers?*

Reviewer 2: Would be good to also talk about the alternatives or how to prevent this scenario

>  * if your *interpreter* is *experimental*, or incapable of running *synchronous* events, the `event` reference might be less useful

Reviewer 3: I think unless we're planning to release/encourage users to release experimental interpreters, we can probably cut this.

Reviewer 2: I agree that, written as is, it can add more confusion. I.e., what should the user do? Are we saying "don't use this feature for experimental interpreters"?

ℹ️ - Please note that if your code runs via *XWorker*, hence in a different thread, there are different caveats and constraints to consider. Please read the [XWorker](#xworker) dedicated section to know more.

> #### The type-env

Reviewer 3: ```suggestion
#### The [type]-env attribute
```

> As `env` as attribute has been mentioned already, it is also possible to confine events to a specific *env* by using the *interpreter* or *custom type* name, and followed by `-env`, for the attribute value, providing the desired *env* to interpret and execute the listener.

Reviewer 3: ```suggestion
Just as the `env` attribute on a \<script\> tag specifies a specific instance of an interpreter to use to run code, it is possible to use the `[type]-env` attribute to specify which instance of an interpreter or custom type should be used to run event code:
```

```html
<script type="micropython">
    def log():
        print(1)
</script>
<!-- note the env value -->
<script type="micropython" env="two">
    # the button will log 2
    def log():
        print(2)
</script>
<!-- note the micropython-env value -->
<button
    micropython-env="two"
    micropython-click="log()"
>
    log
</button>
```

As mentioned before, this will work with `py-env` too, or any custom type defined out there.

> 

Reviewer 3: ```suggestion

It is not currently possible for an element to have register event listeners for two different interpreters of the same type using this syntax.
```
I think this is true?

Author: not sure I am following ... any example you have in mind so I can add it to disambiguate / avoid confusion?

Reviewer 3: For instance, there's no way to have 'click' responders associated to two different envs on the same element, right?

```html
<script type="micropython" env="one">
    def foo():
        print("FOO!")
</script>
<script type="micropython" env="two">
    def bar():
        print("BAR!")
</script>

<!-- 
    There's no way to link this button to functions in both env one and env two, right?
    I.e. Is there a way for a click to run both foo() and bar()? 
    Or for a click to run foo() and a mouseover to run bar()?
-->
<button micropython-env="???" micropython-click="???">Click Me</button>
```

Author: I think there's no use case for that so I am not sure this is a real-world concern ... *env* is an **advanced** feature and whoever uses it would realize it's never desired to have other envs meshed up with listeners as it completely defeat the purpose of an env ... you bootstrap two different worlds, why would you want to trigger listeners all over the place?

## XWorker

Whenever computing relatively expensive stuff, such as a *matplot* image, or literally anything else that would take more than let's say 100ms to answer, running your *interpreter* of choice within a [Web Worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) is likely desirable, so that the main / UI thread won't block users' actions, listeners, or any other computation going on in these days highly dynamic pages.

> However, the native `Worker` utility is incapable of parsing anything that is not *JS* so that we thought that providing a *cross-worker* utility, such as `XWorker` is, would be a great addiction to what *core* offers out of the box.

Reviewer: ```suggestion
However, the native `Worker` utility is incapable of parsing anything that is not *JS* so that we thought that providing a *cross-worker* utility, such as `XWorker` is, would be a great addition to what *core* offers out of the box.
```

> However, the native `Worker` utility is incapable of parsing anything that is not *JS* so that we thought that providing a *cross-worker* utility, such as `XWorker` is, would be a great addiction to what *core* offers out of the box.

Reviewer 3: ```suggestion
@pyscript/core adds a functionality called `XWorker` to all of the interpreters it offers, which works in each language the way `Worker` does in JavaScript.
```

> This reference is a *global* one simply because it has no space in the *JS* global namespace, but also because not all interpreters might provide an utility to define foreign modules out of the *JS* world.

Reviewer 3: ```suggestion
In each Interpreter, `XWorker` is a global reference. 
```

# PyScript Core Documentation

 * [Terminology](#terminology) - what we mean by "_term_" in this document
 * [Bootstrapping core](#bootstrapping-core) - how to enable PyScript Next in your page
 * [How Scripts Work](#how-scripts-work) - how `<script type="...">` works
 * [How Events Work](#how-events-work) - how `<button py-click="...">` works
 * [XWorker](#xworker) - how `XWorker` class and its `xworker` reference work
 * [Custom Scripts](#custom-scripts) - how *custom types* can be defined and used to enrich any core feature

## Terminology

This section goal is to avoid confusion around topics discussed in this document, describing each *term* as exhaustively as possible.

<details>
  <summary><strong>Interpreter</strong></summary>
  <div>

Also commonly referred as *runtime* or *engine*, we consider an **interpreter** any "_piece of software_" able to parse, understand, and ultimately execute, a *Programming Language* through this project.

We also explicitly use that "_piece of software_" as the interpreter name it refers to. We currently bundle references to four interpreters:

>  * [pyodide](https://pyodide.org/en/stable/index.html) is the name of the interpreter that runs likely the most complete version of latest *Python*, enabling dozen official modules at run time, also offering a great *JS* integration in its core

Reviewer 2: dozen --> the full standard library and support installing dozens of modules

 * [micropython](https://micropython.org/) is the name of the interpreter that runs a small subset of the *Python* standard library and is optimized to run in constrained environments such as *Mobile* phones, or even *Desktop*, thanks to its tiny size and an extremely fast bootstrap
 * [wasmoon](https://github.com/ceifa/wasmoon) is the name of the interpreter that runs *Lua* on the browser and that, among the previous two interpreters, is fully compatible with all core features

>  * [ruby-wasm-wasi](https://github.com/ruby/ruby.wasm) is the name of the (currently *experimental*) interpreter that adds *Ruby* to the list of programming languages currently supported

Reviewer 2: Isn't wasmoon also "experimental"? Or do you mean the interpreter itself is experimental?

Author: wasmoon can do everything Pyodide or MicroPython can do ... why do you think it's experimental?

</script>
<div id="output">
    <!-- will show "Hello" once the script executes -->
</div>
```

ℹ️ - Please note that if no `target` attribute is specified, the *script* will automatically create a "_companion element_" when the `target` property/field is accessed for the very first time:

```html
<script type="micropython">
    from js import document

    # will create a <script-micropython> element appended
    # right after the currently executing script
    document.currentScript.target.textContent = "Hello";
</script>
<!--
    created during previous code execution

    <script-micropython>Hello</script-micropython>
-->
```

  </div>
</details>

<details>
  <summary><strong>Env</strong></summary>
  <div>

ℹ️ - This is an **advanced feature** that is worth describing but usually it is not needed for most common use cases.

Mostly due its terseness that plays nicely as attribute's suffix, among its commonly understood meaning, we consider an *env* an identifier that guarantee the used *interpreter* would always be the same and no other interpreters, even if they point at very same project, could interfere with globals, behavior, or what's not.

> In few words, every single *env* would spawn a new interpreter dedicated to such env, and global variables defined elsewhere will not affect this "_environment_" and vice-versa, an *env* cannot dictate what will happen to other interpreters.

Reviewer 2: I find this description a bit confusing... In general, `env` refers to `enviroment`. In this case, it's something like this:

* an environment is an isolated interpreter instance with a specific configuration (packages and files installed into it)
* these environments are identified by "names"
* the `env` attribute on a script allows users to reference a specific environment by name

NOTE:  (Don't want to be picky... just PTSD about communicating the concept of `Enviroments` in the context of Python package managers (`conda`, `pip`, etc..)

Author: `env` here defines a unique environment ... every *env* is a different pyodide or a different MicroPython ... I don't personally know any other terminology to define a runtime/interpreter environment, and *env* plays very well with attributes as it's also very short.

The package manager for the *env* import stuff *only* for that *env*, so which part is not clear or you think should be improved as description? It is, in fact, a unique environment for *micropip* as it won't import anything else in any other pyodide out there that is *not the same env* ... am I misunderstanding your comment or did I really mess up docs in here?

```html
<!-- default env per each interpreter -->
<script type="micropython">
    shared = True
</script>
<script type="micropython">
    # prints True - shared is global
    print(shared)
</script>

<!-- dedicated interpreter -->
<script type="micropython" env="my-project-env">
    # throws an error - shared doesn't exist
    print(shared)
</script>
```

> ℹ️ - Please note if the interpreter takes 1 second to bootstrap, multiple *environments* will take *that* second multiplied by the number of different environments, which is why this feature is considered for **advanced** use cases only and it should be discouraged as generic practice.

Reviewer 2: It'd be good to add how those seconds add up here (how much can run in parallel and how much is serial)

Author: every *env* is a different pyodide, the only removed time with multiple envs is the time to download the module ... let's say 100ms ... if `loadPyodide` takes 1 second to bootstrap, there's no parallel in the same thread, every second will sum up to the amount of envs, the main thread is one, has no parallel work there with WASM modules running in the very same thread.

</script>
```

Please keep reading this document to understand how to use those utilities or how to have other *Pogramming Languages* enabled in your page via `<script>` elements.

## How Scripts Work

The [&lt;script&gt; element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script) has at least these extremely important peculiarities compared to any other element defined by the [HTML Standard](https://html.spec.whatwg.org/multipage/):

 * its only purpose is to contain *data blocks*, meaning that browsers will never try to parse its content as generic *HTML* (and browsers will completely ignore either its content or its attributes, including the `src`, when its [type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type) is not known)
 * its completely unobtrusive when it comes to both *aria* and *layout*, indeed it's one of the few nodes that can be declared almost anywhere without breaking its parent tree (other notable exception would be a comment node)
 * for our specific use case, it already offers attributes that are historically well understood and known, also simplifying somehow the creation of this document

> The long story short is that any `<script type="non-standard-type">` has zero issues with any browser of choice, but it's true that using some specific *custom type* might lead to future issues in case that `type` could have some special meaning for the future of the Web.

Reviewer 2: If this is referring to `Custom Elements` and the fact that inline code is broken, I'd be more explicit here...

Author: no, it refers to the fact that `<script type="python">` or `<script type="json">` might conflict in the future, there's nothing about *custom elements* in core because no *custom element* in the registry is strictly needed.

We encourage everyone to be careful when using this *core* API as we definitively don't want to clash or conflict, by any mean, with what the Web might need or offer in the near to far future, but we're also confident so far our current *types* are more than safe.

### Script Attributes

| name      | example                                       | behavior |
| :-------- | :-------------------------------------------- | :--------|
| async     | `<script type="pyodide" async>`               | The code is evaluated via `runAsync` utility where, if the *interpreter* allows it, top level *await* would be possible, among other *PL* specific asynchronous features.  |
| config    | `<script type="pyodide" config="./cfg.toml">` | The interpreter will load and parse the *JSON* or *TOML* file to configure itself. Please see [currently supported config values](https://docs.pyscript.net/latest/reference/elements/py-config.html#supported-configuration-values) as this is currently based on `<py-config>` features. |

> | env       | `<script type="pyodide" env="brand">`         | Create, if not known yet, a dedicated *environment* for the specified `type`. Please read the [Terminology](#terminology) **env** dedicated details to know more. |

Reviewer 2: Create, if not known yet, a dedicated unique *environment* for the specified `type`. If that environment has already been created, it'll reuse it instead. Please read the [Terminology](#terminology) **env** dedicated details to know more.

```html
<script type="micropython">
    # XWorker is globally defined
    print(XWorker != None)
</script>
```

Please read the [XWorker](#xworker) dedicated section to know more.

  </div>
</details>

## How Events Work

Inspired by the current [HTML Standard](https://html.spec.whatwg.org/multipage/webappapis.html#event-handlers):

> the event handler is exposed through a name, which is a string that always starts with "_on_" and is followed by the name of the event for which the handler is intended.

We took a similar approach, replacing that `on` prefix with whatever *interpreter* or *custom type* is available on the page, plus a *dash* `-` to avoid clashing with standards:

```html
<script type="micropython">
    def print_type(event, double):
        # logs "click 4"
        print(f"{event.type} {double(2)}")
</script>
<button micropython-click="print_type(event, lambda x: x * 2)">
    print type
</button>

> ```

Reviewer 2: Comment here mostly about future..... I think we'll have to work that muscle a lit bit because (I think!!!!) this will be ambiguous. What happens if you have 2 `micropython` enviroments? In that sense we'd need a way to reference an interpreter by type and/or by name... or something like that. 

Not proposing a solution, just highlighting a possible corner case that we need to consider for the future

| type      | `XWorker('./file.py', type='pyodide')`                   | Define the *interpreter* to use with this worker which is, by default, the same one used within the running code. Please read the [Terminology](#terminology) **interpreter** dedicated details to know more. |
| version   | `XWorker('./file.py', type='pyodide', version='0.23.2')` | Allow the usage of a specific version where, if numeric, must be available through the project *CDN* used by *core* but if specified as fully qualified *URL*, allows usage of any interpreter's version: `<script type="pyodide" version="http://localhost:8080/pyodide.local.mjs">` |

The returning *JS* reference to any `XWorker(...)` call is literally a `Worker` instance that, among its default API, have the extra following feature:

| name      | example                            | behavior |
| :-------- | :--------------------------------- | :--------|
| sync      | `sync = XWorker('./file.py').sync` | Allows exposure of callbacks that can be run synchronously from the worker file, even if the defined callback is *asynchronous*. This property is also available in the global `xworker` reference. |

```python

> sync = XWorker('./file.py').sync

Reviewer 2: Think that using `sync` as variable name here can be more confusing to the end user.... it can become confusing in regards of what's part of the API and what's just the variable name

```python
print(xworker.window.document.title)

xworker.window.document.body.append("Hello Main Thread")

xworker.window.setTimeout(print, 100, "timers too")
```

ℹ️ - Please note that even if non blocking, if too many operations are orchestrated from a *worker*, instead of the *main* thread, the overall performance might still be slower due the communication channel and all the primitives involved in the synchronization process. Feel free to use the `window` feature as a great enabler for unthinkable or quick solutions but keep in mind it is still an indirection.

#### The `sync` utility

This helper does not interfere with the global context but it still ensure a function can be exposed form *main* and be used from *thread* and/or vice-versa.

```python
# main
def alert_user(message):
    import js
    js.alert(message)

w = XWorker('./file.py')
# expose the function to the thread
w.sync.alert_user = alert_user

# thread
if condition == None:
    xworker.sync.alert_user('something wrong!')
```

> ## Custom Scripts

Reviewer 2: Think the name "Scripts" here is a bit confusing (overloaded term).

Author: it's about having custom `<script type="...">` tags ... any suggestion?

| io                        | `wrap.io`                                     | Allow to lazily define different `stdout` or `stderr` via the running *interpreter*. This `io` field can be lazily defined and restored back for any element currently running the code. |
| config                    | `wrap.config`                                 | It is the resolved *JSON* config and it is an own clone per each element running the code, usable also as "_state_" reference for the specific element, as changing it at run time will never affect any other element. |
| run                       | `wrap.run(code)`                              | It abstracts away the need to know the exact method name used to run code synchronously, whenever the *interpreter* allows such operation, facilitating future migrations from an interpreter to another. |
| runAsync                  | `wrap.run(code)`                              | It abstracts away the need to know the exact method name used to run code asynchronously, whenever the *interpreter* allows such operation, facilitating future migrations from an interpreter to another. |

> This is the `wrap` mentioned with most hooks and initializers previously described, and we're more than happy to learn if we are not passing along some extra helper.

Reviewer 2: All of the above is absolutely amazing! I think that adding an example of these would go a looooooong way to help users understand how they can be used (Yes, these are advanced feature but, still, would be great)

### The io helper

> 

Reviewer 2: missing description here
